GMG: clean up cell data storage

tjhei opened the pull request:
We now store more information than just the viscosity in the matrix-free
framework to be able to apply the Stokes and GMG operators. Instead of
passing a large number of pointers to Table<>, pass a struct with all
necessary information.

tjhei commented:
@zjiaqi2018 can you please review this refactoring? Note that you now already have the variables for the derivatives on each level.

zjiaqi2018 commented on source/simulator/stokes_matrix_free.cc:
```diff
@@ -2877,23 +2862,19 @@ namespace aspect
   template <int dim, int velocity_degree>
-  const Table<2, VectorizedArray<double>> &
-                                       StokesMatrixFreeHandlerImplementation<dim, velocity_degree>::get_active_viscosity_table() const
+  std::size_t
```
Starting from Line 2808, we need one more empty line between functions.

zjiaqi2018 commented on source/simulator/stokes_matrix_free.cc:
```diff
@@ -660,6 +660,34 @@ namespace aspect
+    OperatorCellData<dim,number>::clear()
+    {
+      enable_newton_derivatives = false;
+      // TODO: use Table::clear() once implemented in 10.0.pre
```
Just got approved, should be able to do it soon!

tjhei replied:
I leave this for a later time, if you don't mind.

zjiaqi2018 replied:
I mean after this PR got merged.

zjiaqi2018 reviewed (commented):
It's in pretty good shape and looks much better now. I only left comments for the documentation and indentation.

tjhei commented:
I think I addressed your comments. Sorry for not fixing the indentation of the comments earlier.

@geodynamics/aspect-developers this is ready for review.

bangerth commented on include/aspect/stokes_matrix_free.h:
```diff
@@ -57,6 +57,73 @@ namespace aspect
+       * otherwise, it is 1.
+       */
+      Table<2, SymmetricTensor<2, dim, VectorizedArray<number>>>
+      viscosity_derivative_wrt_strain_rate_table;
```
Can you name the variable to reflect that it contains this product? Right now it looks like it's just one of the three factors. Maybe something like `Newton_derivative_factor_table` if you don't want to put all of the three factors into the name.

bangerth commented on include/aspect/stokes_matrix_free.h:
```diff
@@ -57,6 +57,73 @@ namespace aspect
+       * Determine an estimate for the memory consumption (in bytes) of this
+       * object.
+       */
+      inline std::size_t
```
```suggestion
      std::size_t
```

bangerth commented on include/aspect/stokes_matrix_free.h:
```diff
@@ -57,6 +57,73 @@ namespace aspect
+      /**
+       * Table which stores viscosity values for each cell.
+       */
+      Table<2, VectorizedArray<number>> viscosity;
```
There are a number of 2d tables in this structure. Would you mind documenting in the class doc what the indices of these tables correspond to?

bangerth commented on include/aspect/stokes_matrix_free.h:
```diff
@@ -182,13 +198,9 @@ namespace aspect
-         * Fills in the viscosity table and sets the value for the pressure scaling constant. The input
-         * @p is_mg_level_data describes whether the viscosity values are defined for a multigrid level
-         * matrix or for the active level matrix.
+         * Pass in a refernce to the problem data.
```
```suggestion
         * Pass in a reference to the problem data.
```

bangerth commented on include/aspect/stokes_matrix_free.h:
```diff
@@ -77,24 +144,9 @@ namespace aspect
-        /**
-         * Fills in the tables of derivatives and strain rate for the Newton method,
-         * and symmetrize the Newton system when @p symmetrize_newton_system is true (i.e., the stabilization is symmetric or SPD).
+         * Pass in a refernce to the problem data.
```
```suggestion
         * Pass in a reference to the problem data.
```

bangerth commented on include/aspect/stokes_matrix_free.h:
```diff
@@ -255,24 +262,9 @@ namespace aspect
-         * Fills in the viscosity table and gives information regarding compressibility. The input
-         * @p is_mg_level_data describes whether the viscosity values are defined for a multigrid level
-         * matrix or for the active level matrix.
+         * Pass in a refernce to the problem data.
```
```suggestion
         * Pass in a reference to the problem data.
```

bangerth commented on source/postprocess/memory_statistics.cc:
```diff
@@ -58,13 +58,12 @@ namespace aspect
         {
-          double mg_transfer_mem = this->get_stokes_matrix_free().get_mg_transfer_A().memory_consumption()
-                                   + this->get_stokes_matrix_free().get_mg_transfer_S().memory_consumption();
+          const double mg_transfer_mem = this->get_stokes_matrix_free().get_mg_transfer_A().memory_consumption()
```
I think this should be a `std::size_type` or similar, not a `double`.

tjhei replied:
We are storing all quantities in memory statistics as doubles and in MB. Maybe because you can see values less than 1 MB? If you want this changed, can we do this in a separate PR?

bangerth replied:
I see. No, not necessary unless you really want!

bangerth commented on source/postprocess/memory_statistics.cc:
```diff
@@ -58,13 +58,12 @@ namespace aspect
-          double visc_table_mem = this->get_stokes_matrix_free().get_active_viscosity_table().memory_consumption()
-                                  + this->get_stokes_matrix_free().get_level_viscosity_tables().memory_consumption();
-          statistics.add_value ("Matrix-free viscosity tables memory consumption (MB) ", visc_table_mem/mb);
+          const double cell_data_mem = this->get_stokes_matrix_free().get_cell_data_memory_consumption();
```
same here

bangerth commented:
Nice cleanups!

tjhei commented:
Updated. I left the memory consumption as doubles in MB, see above.

tjhei commented:
Thank you! @zjiaqi2018 This is now merged but variable names changed.

zjiaqi2018 commented:
Thanks, I will work on the newton gmg preconditioner soon.
